File: services/web-test/frontend/src/lib/sources.ts
/**
 * Fetch-функции для /api/sources/* (web-test backend → catalog).
 *
 * Все ручки провайдер-параметрические: provider передаётся явно в URL, чтобы
 * один и тот же UI-код работал с любым подключённым источником.
 */

const BASE = '/api/sources'

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(`${res.status} ${await res.text()}`)
  return res.json() as Promise<T>
}

// ─── Types ────────────────────────────────────────────────────────────────────

export type ScrapeRunStatus =
  | 'running' | 'ready' | 'applied' | 'discarded' | 'failed'

export type ScrapeRunTotals = {
  new?: number
  updated?: number
  unchanged?: number
  total_slugs?: number
  errors?: number
  applied?: number
}

export type ScrapeRun = {
  id: number
  provider: string
  status: ScrapeRunStatus
  params: Record<string, unknown>
  totals: ScrapeRunTotals
  error_message: string | null
  log_lines: string[]
  started_at: string
  finished_at: string | null
  performed_by: string | null
}

export type ScrapeRunList = { runs: ScrapeRun[]; total: number }

export type ScrapeChangeType = 'new' | 'updated' | 'unchanged'

export type ScrapeItem = {
  id: number
  run_id: number
  slug: string
  payload: Record<string, unknown>
  content_hash: string
  prev_hash: string | null
  change_type: ScrapeChangeType
  field_diffs: Record<string, { before: unknown; after: unknown }> | null
  fetched_at: string
}

export type ScrapeItemList = { items: ScrapeItem[]; total: number }

export type ScrapeRunCreate = {
  max_items?: number | null
  only_year?: number | null
  extra?: Record<string, unknown>
}

export type ScrapeRunApplyRequest = {
  item_ids?: number[]
  change_types?: ScrapeChangeType[]
  performed_by?: string
}

export type MatchWeights = { ru: number; en: number; alias: number }

export type MatchParams = {
  threshold: number
  prefer_external_id: boolean
  weights: MatchWeights
}

export type MatchProfile = {
  id: number
  provider: string
  name: string
  params: MatchParams
  is_default: boolean
  updated_at: string
}

// ─── Detection runs ───────────────────────────────────────────────────────────

export const startSourceRun = (provider: string, body: ScrapeRunCreate) =>
  fetch(`${BASE}/${provider}/runs`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  }).then(r => json<ScrapeRun>(r))

export const fetchSourceRuns = (provider: string, limit = 20, offset = 0) => {
  const u = new URL(`${BASE}/${provider}/runs`, window.location.origin)
  u.searchParams.set('limit', String(limit))
  u.searchParams.set('offset', String(offset))
  return fetch(u.toString().replace(window.location.origin, ''))
    .then(r => json<ScrapeRunList>(r))
}

export const fetchSourceRun = (provider: string, runId: number) =>
  fetch(`${BASE}/${provider}/runs/${runId}`).then(r => json<ScrapeRun>(r))

export const fetchSourceRunItems = (
  provider: string,
  runId: number,
  opts: { changeType?: ScrapeChangeType; search?: string; limit?: number; offset?: number } = {},
) => {
  const u = new URL(
    `${BASE}/${provider}/runs/${runId}/items`,
    window.location.origin,
  )
  if (opts.changeType) u.searchParams.set('change_type', opts.changeType)
  if (opts.search) u.searchParams.set('search', opts.search)
  u.searchParams.set('limit', String(opts.limit ?? 50))
  u.searchParams.set('offset', String(opts.offset ?? 0))
  return fetch(u.toString().replace(window.location.origin, ''))
    .then(r => json<ScrapeItemList>(r))
}

export const applySourceRun = (
  provider: string, runId: number, body: ScrapeRunApplyRequest,
) =>
  fetch(`${BASE}/${provider}/runs/${runId}/apply`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  }).then(r => json<{ run_id: number; applied: number }>(r))

export const discardSourceRun = (provider: string, runId: number) =>
  fetch(`${BASE}/${provider}/runs/${runId}/discard`, {
    method: 'POST',
  }).then(r => json<{ run_id: number; status: string }>(r))

// ─── Match profiles ───────────────────────────────────────────────────────────

export const fetchMatchProfiles = (provider: string) =>
  fetch(`${BASE}/${provider}/match-profiles`).then(r => json<MatchProfile[]>(r))

export const upsertMatchProfile = (
  provider: string,
  body: { name: string; params: MatchParams; is_default?: boolean },
) =>
  fetch(`${BASE}/${provider}/match-profiles`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(body),
  }).then(r => json<MatchProfile>(r))

export const deleteMatchProfile = async (provider: string, profileId: number) => {
  const res = await fetch(`${BASE}/${provider}/match-profiles/${profileId}`, {
    method: 'DELETE',
  })
  if (!res.ok && res.status !== 204) throw new Error(`${res.status} ${await res.text()}`)
}

// ─── Promotion candidates с MatchParams ───────────────────────────────────────

export const fetchPromotionCandidatesWithParams = (
  provider: string,
  rawId: number,
  params: MatchParams,
  limit = 5,
) => {
  const u = new URL(
    `${BASE}/${provider}/promotion/${rawId}/candidates`,
    window.location.origin,
  )
  u.searchParams.set('threshold', String(params.threshold))
  u.searchParams.set('limit', String(limit))
  u.searchParams.set('prefer_external_id', String(params.prefer_external_id))
  u.searchParams.set('weight_ru', String(params.weights.ru))
  u.searchParams.set('weight_en', String(params.weights.en))
  u.searchParams.set('weight_alias', String(params.weights.alias))
  return fetch(u.toString().replace(window.location.origin, '')).then(r => json(r))
}

export const DEFAULT_MATCH_PARAMS: MatchParams = {
  threshold: 0.5,
  prefer_external_id: false,
  weights: { ru: 1.0, en: 1.0, alias: 1.0 },
}

File: services/web-test/app/api/sources.py
"""Прокси к /sources/* эндпоинтам catalog'а.

Всё под /api/sources. Контракт идентичен upstream'у — это тонкий форвард,
чтобы фронту не нужно было ходить cross-origin и держать `X-API-Key`.

Покрывает:
  - detection runs: запуск/листинг/просмотр/apply/discard
  - match profiles: CRUD сохранённых конфигов матчинга
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query

from app.catalog_client import CatalogClient, CatalogServiceError
from app.deps import get_catalog_client

router = APIRouter(prefix="/sources", tags=["sources"])


# ─── Detection runs ───────────────────────────────────────────────────────────


@router.post("/{provider}/runs")
async def start_run(
    provider: str,
    payload: dict[str, Any],
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    """Запустить сухой прогон. Возвращается run сразу с status='running' —
    UI начинает polling до status='ready'/'failed'."""
    try:
        return await client.start_source_run(provider, payload)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.get("/{provider}/runs")
async def list_runs(
    provider: str,
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.list_source_runs(provider, limit=limit, offset=offset)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.get("/{provider}/runs/{run_id}")
async def get_run(
    provider: str,
    run_id: int,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.get_source_run(provider, run_id)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.get("/{provider}/runs/{run_id}/items")
async def list_run_items(
    provider: str,
    run_id: int,
    change_type: str | None = Query(None),
    search: str | None = Query(None),
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.list_source_run_items(
            provider,
            run_id,
            change_type=change_type,
            search=search,
            limit=limit,
            offset=offset,
        )
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.post("/{provider}/runs/{run_id}/apply")
async def apply_run(
    provider: str,
    run_id: int,
    payload: dict[str, Any],
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.apply_source_run(provider, run_id, payload)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.post("/{provider}/runs/{run_id}/discard")
async def discard_run(
    provider: str,
    run_id: int,
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.discard_source_run(provider, run_id)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


# ─── Match profiles ───────────────────────────────────────────────────────────


@router.get("/{provider}/match-profiles")
async def list_match_profiles(
    provider: str,
    client: CatalogClient = Depends(get_catalog_client),
) -> list[dict[str, Any]]:
    try:
        return await client.list_match_profiles(provider)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.post("/{provider}/match-profiles")
async def upsert_match_profile(
    provider: str,
    payload: dict[str, Any],
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    try:
        return await client.upsert_match_profile(provider, payload)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


@router.delete("/{provider}/match-profiles/{profile_id}", status_code=204)
async def delete_match_profile(
    provider: str,
    profile_id: int,
    client: CatalogClient = Depends(get_catalog_client),
) -> None:
    try:
        await client.delete_match_profile(provider, profile_id)
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e


# ─── Promotion candidates с MatchParams ───────────────────────────────────────


@router.get("/{provider}/promotion/{raw_id}/candidates")
async def promotion_candidates_with_params(
    provider: str,
    raw_id: int,
    threshold: float = Query(0.5, ge=0.0, le=1.0),
    limit: int = Query(5, ge=1, le=20),
    prefer_external_id: bool = Query(False),
    weight_ru: float = Query(1.0, ge=0.0, le=2.0),
    weight_en: float = Query(1.0, ge=0.0, le=2.0),
    weight_alias: float = Query(1.0, ge=0.0, le=2.0),
    client: CatalogClient = Depends(get_catalog_client),
) -> dict[str, Any]:
    """Кандидаты с настраиваемыми параметрами матчинга.

    Дублирует /api/catalog/promotion/{provider}/{raw_id}/candidates, но
    плоско прокидывает MatchParams в query. UI MatchParamsForm зовёт
    эту ручку.
    """
    try:
        return await client.promotion_candidates_with_params(
            provider,
            raw_id,
            threshold=threshold,
            limit=limit,
            prefer_external_id=prefer_external_id,
            weight_ru=weight_ru,
            weight_en=weight_en,
            weight_alias=weight_alias,
        )
    except CatalogServiceError as e:
        raise HTTPException(status_code=e.status_code, detail=e.detail) from e
